### PR TITLE
Make use of `used` in unittest

### PR DESCRIPTION
**Pull request by bluenote10**
The addition of the `used` pragma is really great! The `unittest` module is the place where I wanted it most.

Maybe I should hack the testament specs to allow for a `numExpectedWarnings: 0` specification, so that the effect of the `used` pragma can be tested?

### CODE REVIEW

**Comment by Araq** [comment]
In the spec ``nimout: ""`` should do.

**Comment by bluenote10** [comment]
The problem with `nimout` is that it only checks that the given string is part of the compiler output, so `nimout: ""` would always be true.

I realized that `numExpectedWarnings` does not work either, because `XDeclaredButNotUsed` is a hint and not a warning. For a test where we want to test that there is exactly one `XDeclaredButNotUsed` the output would look like:

```
Hint: used config file '/home/fabian/bin/nim-repo/config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: tused [Processing]
tused.nim(15, 8) Hint: 'tused.echoSub(a: int, b: int)' is declared but not used [XDeclaredButNotUsed]
Hint:  [Link]
Hint: operation successful (10736 lines compiled; 0.113 sec total; 17.938MiB peekmem; Debug Build) [SuccessX]
```

Overall, we want a solution which is easy to maintain, i.e., a completely unrelated change affecting compiler output should not require to adjust the tests. Options:

- Checking for the number of hints => fragile
- Checking for full compiler output => fragile
- Checking for compiler output context (i.e., matching would include just the last hint before and after) => a bit better.
- Having a static echo at the beginning and end to define a clear, explicit context.

I went for the last solution: Does not require extending the test specs and due to the explicitly define context, the matching should be pretty good in terms of maintainability. What do you think?

File: tests/pragmas/tused.nim [hunk 1]
```diff
@@ -1,13 +1,35 @@
...
+tused.nim(15, 8) Hint: 'tused.echoSub(a: int, b: int)' is declared but not used [XDeclaredButNotUsed]
+compile end'''
+  output: "8\n8"
 """
 
-template implementArithOps(T) =
+static:
+  echo "compile start"
+
+template implementArithOpsOld(T) =
```
Comment (bluenote10) [inline tests/pragmas/tused.nim]: I have modified the test to check that `{.used.}` actually makes a difference.

**Comment by bluenote10** [comment]
Oh, the Windows builds are failing, and actually, what I just tried will probably not work either. Any ideas where these [double newlines](https://ci.appveyor.com/project/Araq/nim/build/225#L2772) in the compiler output are coming from?

**Comment by bluenote10** [comment]
I solved the problem by simply filtering the nimout for empty lines. I assume that the actual problem lies in `osproc`. From the output I liked above, it looks like the problem is the mixing of stdout and stderr: Things that go to stdout (the static echo) only get one newline, things that go to stderr (the hints) get two newlines (but only on Windows). Probably a very low priority problem, so I hope that filtering out empty lines is sufficient for now.

**Comment by Araq** [comment]
> Probably a very low priority problem, so I hope that filtering out empty lines is sufficient for now.

No, sorry. This should be fixed properly, in the past hacks/workarounds in testament did really bite us.

**Comment by bluenote10** [comment]
Agreed that is definitely better. And I think I know what's going wrong, actually worth a bugfix anyway. Will file a PR later/tomorrow.

File: tests/testament/tester.nim [hunk 2]
```diff
@@ -76,7 +76,8 @@ proc callCompiler(cmdTemplate, filename, options: string,
   var x = newStringOfCap(120)
   result.nimout = ""
   while outp.readLine(x.TaintedString) or running(p):
-    result.nimout.add(x & "\n")
+    if x.strip != "":
```
Comment (Araq) [inline tests/testament/tester.nim]: remove this workaround now please and it's fine.

Comment (bluenote10) [inline tests/testament/tester.nim]: done (rebased & reverted)

**Pull request merged**